### PR TITLE
[DDCE-507] Updated error messages on /your-details

### PR DESCRIPTION
**Pull request by DannyPennington97**
# DDCE-507

## Content change

Updating error messages on /your-details
Added new unit test for validation failures
No dependency updates required as they have just been done
Link to acceptance test PR: https://github.com/hmrc/lisa-api-tests/pull/97

~~Email regex was not required to be implemented as we are using the "email" mapping in the form model (see [here](https://html.spec.whatwg.org/multipage/input.html#e-mail-state-(type%3Demail)) for more info)~~
Have implemented custom email regex similar to other services because the play default mapping for email allows things like "test@test" (missing domain) which we can't have

## PR Suggestions
- Have you looked at the JIRA story to make sure all Acceptance Criteria has been met?
- Have you done a visual check of the changes?
- Is there any still commented or unused code? Lingering printlns?
- Have things been implemented in a complicated way?
- Are the test still over the coverage threshold?
- Does the compiler throw a lot of warnings? 
- Have methods and tests been named clearly?
- Were there any changes in the config? Do changes need to be made in app-config-???

## Checklist PR Raiser
 - [x]  I've made every effort to commit high quality, clean code and I have executed relevant static analyses to be sure
 - [x]  I've included appropriate tests with any code I've added (Unit, Integration, Acceptance etc.)
 - [x]  I've executed the acceptance test pack locally to ensure there are no functional regressions (Performance tests where relevant)
 - [x]  I've squashed my commits - including the JIRA issue number in the commit message
 - [x]  I've run a dependency check to ensure all dependencies are up to date

## Checklist PR Reviewer
 - [x]  I've made every effort to commit high quality, clean code and I have executed relevant static analyses to be sure
 - [x]  I've included appropriate tests with any code I've added (Unit, Integration, Acceptance etc.)
 - [x]  I've executed the acceptance test pack locally to ensure there are no functional regressions (Performance tests where relevant)
 - [x]  I've squashed my commits - including the JIRA issue number in the commit message
 - [ ]  I've run a dependency check to ensure all dependencies are up to date

### CODE REVIEW

File: conf/messages [hunk 1]
```diff
@@ -161,19 +161,19 @@ error.isaProviderRefNumberRequired=Enter your ISA manager reference
 error.isaProviderRefNumberPattern= Your company’s ISA manager reference must only include the letter z and numbers 0-9
 
 error.firstNameRequired=Enter your first name
-error.firstNameLength=Enter only the first 35 characters of your first name
-error.firstNamePattern=First name can only include letters a to z, 0 to 9, ampersands, apostrophes, back slashes, full stops, hyphens and spaces
+error.firstNameLength=First name must be 35 characters or less
+error.firstNamePattern=First name must only include letters a to z, hyphens, spaces and apostrophes
```
Comment (sean-corrigan) [inline conf/messages:165]: "0 to 9" , this bit has been taken out of the message but the field still takes numbers

File: conf/messages [hunk 2]
```diff
@@ -161,19 +161,19 @@ error.isaProviderRefNumberRequired=Enter your ISA manager reference
...
 
 error.lastNameRequired=Enter your last name
-error.lastNameLength=Enter only the first 35 characters of your last name
-error.lastNamePattern=Last name can only include letters a to z, 0 to 9, ampersands, apostrophes, back slashes, full stops, hyphens and spaces
+error.lastNameLength=Last name must be 35 characters or less
+error.lastNamePattern=Last name must only include letters a to z, hyphens, spaces and apostrophes
```
Comment (sean-corrigan) [inline conf/messages:169]: "0 to 9" , this bit has been taken out of the message but the field still takes numbers (same as first name)

File: conf/messages [hunk 3]
```diff
@@ -161,19 +161,19 @@ error.isaProviderRefNumberRequired=Enter your ISA manager reference
...
 
 error.roleRequired=Enter your job title
-error.roleLength=Enter only the first 30 characters of your job title
-error.rolePattern=Job title can only include letters a to z, 0 to 9, ampersands, apostrophes, forward slashes, full stops, hyphens and spaces
+error.roleLength=Job title must be 30 characters or less
+error.rolePattern=Job title must only include letters a to z, hyphens, spaces and apostrophes
```
Comment (sean-corrigan) [inline conf/messages:173]: "0 to 9" , this bit has been taken out of the message but the field still takes numbers (same as first name)

File: conf/messages (hunk 1 above)
Comment (DannyPennington97) [reply conf/messages:165]: Ah yep, will update the regex then

File: app/models/YourDetails.scala [hunk 4]
```diff
@@ -36,24 +36,25 @@ object YourDetails {
...
+        .verifying("error.rolePattern", i => i.isEmpty || i.getOrElse("").length > 30 || i.getOrElse("").matches("""^[A-Za-z \-']{1,30}$""")),
 
       "phone" -> optional(text)
         .verifying("error.phoneRequired", _.isDefined)
         .verifying("error.phonePattern", i => i.isEmpty || i.getOrElse("").matches("""^[A-Z0-9 \)\/\(\*\#\-\+]{1,24}$""")),
 
-      "email" -> optional(email)
+      "email" -> optional(text)
         .verifying("error.emailRequired", _.isDefined)
+        .verifying("error.email", i => i.isEmpty || i.getOrElse("").matches("""^[a-zA-Z0-9_+&*-](?:\.[a-zA-Z0-9_+&*-])*@(?:[a-zA-Z0-9-].)[a-zA-Z]{2,7}$""")),
```
Comment (sean-corrigan) [inline app/models/YourDetails.scala:57]: I know this regex was in the ticket but i put a couple of emails in and all fail validation, also causing the ATs to fail.

Comment (DannyPennington97) [reply app/models/YourDetails.scala:57]: Hmm yeah, this regex doesn't seem to work at all...
Just testing an updated one now

**Comment by DannyPennington97** [comment]
Updated to address comments

File: app/models/YourDetails.scala [hunk 5]
```diff
@@ -36,24 +36,25 @@ object YourDetails {
...
+        .verifying("error.rolePattern", i => i.isEmpty || i.getOrElse("").length > 30 || i.getOrElse("").matches("""^[A-Za-z \-']{1,30}$""")),
 
       "phone" -> optional(text)
         .verifying("error.phoneRequired", _.isDefined)
         .verifying("error.phonePattern", i => i.isEmpty || i.getOrElse("").matches("""^[A-Z0-9 \)\/\(\*\#\-\+]{1,24}$""")),
 
-      "email" -> optional(email)
+      "email" -> optional(text)
         .verifying("error.emailRequired", _.isDefined)
+        .verifying("error.email", i => i.isEmpty || i.getOrElse("").matches("""(?:[A-Za-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[A-Za-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[A-Za-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9]))\.){3}(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9])|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])"""))
```
Comment (sean-corrigan) [inline app/models/YourDetails.scala:57]: beast of a regex, where did that come from 😎 (works just curious lol)

Comment (DannyPennington97) [reply app/models/YourDetails.scala:57]: Yeah, it's a long one! I got it from http://emailregex.com/ but had to add a bit to allow for capitals before the "@"

**Pull request merged**